scipy/optimize/optimize.py

```diff
@@ -716,6 +729,8 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
         else:
             maxfun = np.inf
 
+    sim = np.clip(sim, lower_bound, upper_bound)
```

Comment:
does all the clipping add a lot of overhead to the minimisation?

Reply:
Good point. I checked the benchmark results on CircleCI. 
I'm not sure that VM performance of Circle CI is stable or not, but it seems that this branch is almost 2 times slower than master branch.

```
Desktop $ cat master.txt | grep nelder
                 rosenbrock_slow      nelder-mead     mean_nfev            310.2
                 rosenbrock_slow      nelder-mead     mean_time     0.055134820938110354
                rosenbrock_nograd     nelder-mead     mean_nfev            310.2
                rosenbrock_nograd     nelder-mead     mean_time     0.015512728691101074
                    rosenbrock        nelder-mead     mean_nfev            310.2
                    rosenbrock        nelder-mead     mean_time     0.012282705307006836
                 rosenbrock_tight     nelder-mead     mean_nfev            407.7
                 rosenbrock_tight     nelder-mead     mean_time     0.016884851455688476
                 simple_quadratic     nelder-mead     mean_nfev            152.8
                 simple_quadratic     nelder-mead     mean_time     0.005493426322937011
               asymmetric_quadratic   nelder-mead     mean_nfev            162.9
               asymmetric_quadratic   nelder-mead     mean_time     0.006136679649353027
                      sin_1d          nelder-mead     mean_nfev             36.8
                      sin_1d          nelder-mead     mean_time     0.002183842658996582
                      booth           nelder-mead     mean_nfev            122.9
                      booth           nelder-mead     mean_time    0.0049095630645751955
                      beale           nelder-mead     mean_nfev            141.1
                      beale           nelder-mead     mean_time     0.005901885032653808
                        LJ            nelder-mead     mean_nfev            2281.8
                        LJ            nelder-mead     mean_time      0.2883703947067261
Desktop $ cat PR13515.txt | grep nelder
                 rosenbrock_slow      nelder-mead     mean_nfev            310.2
                 rosenbrock_slow      nelder-mead     mean_time     0.09497151374816895
                rosenbrock_nograd     nelder-mead     mean_nfev            310.2
                rosenbrock_nograd     nelder-mead     mean_time     0.037506628036499026
                    rosenbrock        nelder-mead     mean_nfev            310.2
                    rosenbrock        nelder-mead     mean_time     0.03789718151092529
                 rosenbrock_tight     nelder-mead     mean_nfev            407.7
                 rosenbrock_tight     nelder-mead     mean_time     0.05336172580718994
                 simple_quadratic     nelder-mead     mean_nfev            152.8
                 simple_quadratic     nelder-mead     mean_time     0.015970468521118164
               asymmetric_quadratic   nelder-mead     mean_nfev            162.9
               asymmetric_quadratic   nelder-mead     mean_time     0.017650723457336426
                      sin_1d          nelder-mead     mean_nfev             36.8
                      sin_1d          nelder-mead     mean_time     0.005703425407409668
                      booth           nelder-mead     mean_nfev            122.9
                      booth           nelder-mead     mean_time     0.015587162971496583
                      beale           nelder-mead     mean_nfev            141.1
                      beale           nelder-mead     mean_time     0.020792436599731446
                        LJ            nelder-mead     mean_nfev            2281.8
                        LJ            nelder-mead     mean_time      0.5842879295349122
``` 

One idea  is not clipping when `bounds` is not provided like:

```
if lower_bound != None:
    sim = np.clip(sim, lower_bound, upper_bound)
```
I think we can do the same thing for all `np.clip`.
What do you think?

Reply:
I added the `if bounds is not None` for all np.clip. This is benchmark result of it. It looks the overhead has gone when the bounds is not provided.
```
Desktop $ cat add_none_guard.txt | grep nelder
                 rosenbrock_slow      nelder-mead     mean_nfev            310.2
                 rosenbrock_slow      nelder-mead     mean_time     0.05221078395843506
                rosenbrock_nograd     nelder-mead     mean_nfev            310.2
                rosenbrock_nograd     nelder-mead     mean_time     0.00959794521331787
                    rosenbrock        nelder-mead     mean_nfev            310.2
                    rosenbrock        nelder-mead     mean_time     0.013217735290527343
                 rosenbrock_tight     nelder-mead     mean_nfev            407.7
                 rosenbrock_tight     nelder-mead     mean_time     0.011948251724243164
                 simple_quadratic     nelder-mead     mean_nfev            152.8
                 simple_quadratic     nelder-mead     mean_time    0.0031580686569213866
               asymmetric_quadratic   nelder-mead     mean_nfev            162.9
               asymmetric_quadratic   nelder-mead     mean_time    0.0034789800643920898
                      sin_1d          nelder-mead     mean_nfev             36.8
                      sin_1d          nelder-mead     mean_time    0.0017694473266601563
                      booth           nelder-mead     mean_nfev            122.9
                      booth           nelder-mead     mean_time    0.0028969287872314454
                      beale           nelder-mead     mean_nfev            141.1
                      beale           nelder-mead     mean_time    0.0033657312393188476
                        LJ            nelder-mead     mean_nfev            2281.8
                        LJ            nelder-mead     mean_time     0.17972052097320557
```